core/src/main/scala/org/apache/spark/executor/ProcfsBasedSystems.scala

```diff
@@ -0,0 +1,274 @@
+// project.
+private[spark] class ProcfsBasedSystems(procfsDir: String = "/proc/") extends Logging {
+  val procfsStatFile = "stat"
+  var pageSize: Long = computePageSize()
+  var isAvailable: Boolean = isProcfsAvailable
+  private val pid: Int = computePid()
+  private val ptree: scala.collection.mutable.Map[ Int, Set[Int]] =
+    scala.collection.mutable.Map[ Int, Set[Int]]()
+
+  var allMetrics: ProcfsBasedSystemsMetrics = ProcfsBasedSystemsMetrics(0, 0, 0, 0, 0, 0)
+  private var latestJVMVmemTotal: Long = 0
+  private var latestJVMRSSTotal: Long = 0
+  private var latestPythonVmemTotal: Long = 0
+  private var latestPythonRSSTotal: Long = 0
+  private var latestOtherVmemTotal: Long = 0
+  private var latestOtherRSSTotal: Long = 0
+
+  computeProcessTree()
+
+  private def isProcfsAvailable: Boolean = {
+    val testing = sys.env.contains("SPARK_TESTING") || sys.props.contains("spark.testing")
+    if (testing) {
+      return true
+    }
+    try {
+      if (!Files.exists(Paths.get(procfsDir))) {
+        return false
+      }
+    }
+    catch {
+      case f: FileNotFoundException => return false
+    }
+    val shouldLogStageExecutorMetrics =
+      SparkEnv.get.conf.get(config.EVENT_LOG_STAGE_EXECUTOR_METRICS)
+    val shouldLogStageExecutorProcessTreeMetrics =
+      SparkEnv.get.conf.get(config.EVENT_LOG_PROCESS_TREE_METRICS)
+    shouldLogStageExecutorProcessTreeMetrics && shouldLogStageExecutorMetrics
+  }
+
+  private def computePid(): Int = {
+    if (!isAvailable) {
+      return -1;
+    }
+    try {
+      // This can be simplified in java9:
+      // https://docs.oracle.com/javase/9/docs/api/java/lang/ProcessHandle.html
+      val cmd = Array("bash", "-c", "echo $PPID")
+      val length = 10
+      val out: Array[Byte] = Array.fill[Byte](length)(0)
+      Runtime.getRuntime.exec(cmd).getInputStream.read(out)
+      val pid = Integer.parseInt(new String(out, "UTF-8").trim)
+      return pid;
+    }
+    catch {
+      case e: IOException => logDebug("IO Exception when trying to compute process tree." +
+        " As a result reporting of ProcessTree metrics is stopped", e)
+        isAvailable = false
+        return -1
+      case t: Throwable => logDebug("Some exception occurred when trying to" +
+        " compute process tree. As a result reporting of ProcessTree metrics is stopped", t)
+        isAvailable = false
+        return -1
+    }
+  }
+
+  private def computePageSize(): Long = {
+    val testing = sys.env.contains("SPARK_TESTING") || sys.props.contains("spark.testing")
+    if (testing) {
+      return 0;
+    }
+    val cmd = Array("getconf", "PAGESIZE")
+    val out: Array[Byte] = Array.fill[Byte](10)(0)
+    Runtime.getRuntime.exec(cmd).getInputStream.read(out)
+    return Integer.parseInt(new String(out, "UTF-8").trim)
+  }
+
+  private def computeProcessTree(): Unit = {
+    if (!isAvailable) {
+      return
+    }
+    val queue: Queue[Int] = new Queue[Int]()
```

Comment:
No need to declare type. Can we check everywhere in the patch and verify we're declaring types as infrequently as possible?

Reply:
done